src/testRunner/unittests/tscWatch/forceConsistentCasingInFileNames.ts

```diff
@@ -113,11 +113,154 @@ export const Fragment: unique symbol;
+                        caption: "Prepend a line to moduleA",
+                        change: sys => sys.prependFile(`${windowsStyleRoot}/${projectRootRelative}/a.ts`, `// some comment
+                        `),
+                        timeouts: runQueuedTimeoutCallbacks,
+                    }
+                ],
+            });
+        }
+
+        verifyWindowsStyleRoot("when Windows-style drive root is lowercase", "c:/", "project");
+        verifyWindowsStyleRoot("when Windows-style drive root is uppercase", "C:/", "project");
+
+        function verifyFileSymlink(subScenario: string, diskPath: string, targetPath: string, importedPath: string) {
+            verifyTscWatch({
+                scenario: "forceConsistentCasingInFileNames",
+                subScenario,
+                commandLineArgs: ["--w", "--p", ".", "--explainFiles"],
+                sys: () => {
+                    const moduleA: File = {
+
+                        path: diskPath,
+                        content: `
+export const a = 1;
+export const b = 2;
+`
+                    };
+                    const symlinkA: SymLink = {
+                        path: `${projectRoot}/link.ts`,
+                        symLink: targetPath,
+                    };
+                    const moduleB: File = {
+                        path: `${projectRoot}/b.ts`,
+                        content: `
+import { a } from "${importedPath}";
+import { b } from "./link";
+
+a;b;
+`
+                    };
+                    const tsconfig: File = {
+                        path: `${projectRoot}/tsconfig.json`,
+                        content: JSON.stringify({ compilerOptions: { forceConsistentCasingInFileNames: true } })
+                    };
+                    return createWatchedSystem([moduleA, symlinkA, moduleB, libFile, tsconfig], { currentDirectory: projectRoot });
+                },
+                changes: [
+                    {
+                        caption: "Prepend a line to moduleA",
+                        change: sys => sys.prependFile(diskPath, `// some comment
+                        `),
+                        timeouts: runQueuedTimeoutCallbacks,
+                    }
+                ],
+            });
+        }
+
+        verifyFileSymlink("when both file symlink target and import match disk", `${projectRoot}/XY.ts`, `${projectRoot}/XY.ts`, `./XY`);
+        verifyFileSymlink("when file symlink target matches disk but import does not", `${projectRoot}/XY.ts`, `${projectRoot}/Xy.ts`, `./XY`);
+        verifyFileSymlink("when import matches disk but file symlink target does not", `${projectRoot}/XY.ts`, `${projectRoot}/XY.ts`, `./Xy`);
+        verifyFileSymlink("when import and file symlink target agree but do not match disk", `${projectRoot}/XY.ts`, `${projectRoot}/Xy.ts`, `./Xy`);
+        verifyFileSymlink("when import, file symlink target, and disk are all different", `${projectRoot}/XY.ts`, `${projectRoot}/Xy.ts`, `./yX`);

```

Comment:
I think the testcase missing that we discussed and is important change is when current directory is "c:/temp" but on disk it is "C:/Temp" and all realpath results with new API will differ in casing and give errors vs it wasn't giving earlier ?

Reply:
And may be it is ok for watch scenario but it is something that will impact tsserver and in turn vscode.. vscode keeps track of open files from previous session so not sure if that gets updated if you change the casing and restart vscode from directory but users will not be able to get rid of error easily unless they close all files if the casing is retained across sessions by vscode. You would need to experiment to figure that out.

Reply:
> when current directory is "c:/temp" but on disk it is "C:/Temp"

I attempted to provide test coverage for drive root casing differences [here](https://github.com/microsoft/TypeScript/pull/41292/files#diff-742ed890d46f805d54cbcf33f4a6f2cd286a81ac5df890b0b20636445b464093R162).  It's entirely possible I misunderstood your concerns though - please let me know if there's something else you'd like me to test.

> may be it is ok for watch scenario but it is something that will impact tsserver

From our offline discussion (before the break), I thought I understood that the watch VFS was the best place to add these tests.  It sounds like you may be suggesting adding server tests as well.  If that's the case, can you please point me at an example I can model them on?

> vscode keeps track of open files from previous session

I think you're saying that the paths returned by the server to VS Code might reflect the casing returned by realpathSync, but I'm not sure I understand what consequences you expect from this.  If it's about re-opening files in a new session, I would have guessed VS Code used case-insensitive system calls to open them.  If it's about VS Code passing stale paths back to the server when it starts a fresh instance, I would have guessed they would be re-normalized with an additional realpathSync call.  Can you please elaborate on your concerns?

Reply:
> I attempted to provide test coverage for drive root casing differences here. It's entirely possible I misunderstood your concerns though - please let me know if there's something else you'd like me to test.

That test case also gives incorrect result? [This](https://github.com/microsoft/TypeScript/pull/41292/files#diff-a87a00c391bf2ad3ae44e4560646d9b8725590957f232578e15e0d6230e4477eR36) should have been earlier been error and should be error because on case sensitive file system the import will not resolve, which is what intention of `forceConsistentFileNames` is.. 
The case that it doesnt verify is also when the files on disk and all imports are consistent, but you start with current directory = different casing and build the project. Previously that would not error now it will ? 


> From our offline discussion (before the break), I thought I understood that the watch VFS was the best place to add these tests. It sounds like you may be suggesting adding server tests as well. If that's the case, can you please point me at an example I can model them on?

If the above scenario breaks for watch there is a way to fix those errors by closing tsc and restarting from correct casing directory.  But what happens in vscode is question. vscode keeps list of open files, so is the solution to close all files and open editor again? Does restart work , depending on answers to that we may need coverage for some tests here to lock the expected behaviour for baselines. There are tests for these in `unittests\tsserver\forceConsistentCasingInFileNames.ts`

Reply:
I had the impression that `forceConsistentFileNames` [ignored](https://github.com/microsoft/TypeScript/blob/c456bbd4665e361a7495c3a13fd1ca9bc1265230/src/compiler/program.ts#L2533) drive roots.  Sorry for not including a comment to that effect.

> The case that it doesnt verify is also when the files on disk and all imports are consistent, but you start with current directory = different casing and build the project. Previously that would not error now it will ?

I'm not sure I follow.  Where is the current directory specified in this case?

> But what happens in vscode is question

So you'd just like me to manually verify that VS Code doesn't have obvious breaks if I close a folder, change the casing on disk, and re-open it?

Reply:
> I had the impression that forceConsistentFileNames ignored drive roots. Sorry for not including a comment to that effect.

Yes. But your import doesnt test import "somethingsomething/a" and  "somethingsomething/A" you would get error on either one of those and that was expected behavior which is not the current case.

> I'm not sure I follow. Where is the current directory specified in this case?

Current directory need to be specified as part of the https://github.com/microsoft/TypeScript/pull/41292/files#diff-742ed890d46f805d54cbcf33f4a6f2cd286a81ac5df890b0b20636445b464093R149 . We need test where current directory differs in casing from whats on the disk.

The bigger missing part is import in wrong casing should give error which it wont because the realpath will resolve to casing on the disk and then we will loose that info. But same code in case sensitive file system will give error because import will not be resolved.

Disk file name: a.ts 

import "a" > success
import "A" > will fail on case sensistive file systems... so it should error with forceConsistentFileCasing on case insensitive filesystem which it would not when realpath is invoked as part of this change.